shared/protocol/src/main/java/io/pravega/shared/protocol/netty/CommandEncoder.java

```diff
@@ -65,20 +68,67 @@
 @Slf4j
 public class CommandEncoder extends MessageToByteEncoder<Object> {
     private static final byte[] LENGTH_PLACEHOLDER = new byte[4];
+    private static final int MAX_DATA_SIZE = 4 * 1024 * 1024; // 4MB
     private final Function<Long, AppendBatchSizeTracker> appendTracker;
     private final Map<Map.Entry<String, UUID>, Session> setupSegments = new HashMap<>();
     private final AtomicLong tokenCounter = new AtomicLong(0);
     private String segmentBeingAppendedTo;
     private UUID writerIdPerformingAppends;
     private int currentBlockSize;
     private int bytesLeftInBlock;
+    private final List<Session> pendingWrites = new ArrayList<>();
+    private long totalBytes = 0;
 
     @RequiredArgsConstructor
-    private static final class Session {
+
+    private final class Session {
+        private static final int MAX_BLOCK_SIZE = 1024 * 1024;  // 1MB
+        private static final int MAX_EVENTS = 500;
         private final UUID id;
-        private long lastEventNumber = -1L;
-        private int eventCount;
         private final long requestId;
+        private final ByteBuf data = buffer();
+        private long lastEventNumber = -1L;
+        private int eventCount = 0;
+
+        private void recordAppend(Append append) {
+            lastEventNumber = append.getEventNumber();
+            eventCount += append.getEventCount();
+        }
+
+        private void append(ByteBuf buffer, ByteBuf out) {
+            if (buffer != null && buffer.readableBytes() > 0) {
+                totalBytes += buffer.readableBytes();
+                if (data.readableBytes() == 0) {
+                    pendingWrites.add(this);
```

Comment:
I'm not sure I understand this if block.

Reply:
if channel is not free.. lets append is in progress then we start accumulating/buffer the data from other writers/sessions.